pitrery.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH PITRERY 1 "February 3, 2021"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
pitrery \- Manage point-in-time recovery for PostgreSQL

.SH SYNOPSIS
.B pitrery
.RI [ options ]
.I action
.RI [ args ]

The command needs to be written in a specific order: pitrery options should be
stated first, and the action options should be stated after the action
keywords. This is important since some switches are the same in both lists of
options, but with different meanings.

.SH DESCRIPTION
The purpose of \fBpitrery\fP is to make it easy to manage the archiving of
PostgreSQL WAL segments and automate the tasks of taking base backups and
preparing a point in time recovery operation if or when that is needed.

The \fBpitrery\fP script itself provides six archive maintenance actions,
.BR backup ", " restore ", " purge ", " list ", " check ", and " configure ,
which create a new base backup, prepare the necessary files for
initiating a recovery, clean up old base backup and WAL segments that
are no longer wanted, display the available base backups, and check
and create the configuration, respectively. It is designed to be run
on the same machine as the PostgreSQL cluster, but the base backups
and WAL segments that it operates on can be archived either locally or
on a remote server using an SSH connection.

The \fBpitrery\fP package also provides two helper scripts,
\fBarchive_wal\fP(1) and \fBrestore_wal\fP(1), which can help with the
task of managing the WAL segments that accumulate between base backups,
allowing them to be compressed and/or stored on remote host away from the
database server. It is not required to use these two scripts to use
\fBpitrery\fP, but similar functionality to copy WAL segments to an archive
space and retrieve them on demand, will need to be provided by some other
means if they are not. They can also be called from other scripts if you
require additional functionality that they don't provide. The main advantage
they offer is keeping all needed configuration for backup and restore tasks
coherently in a single location.


.SH OPTIONS
The following options are available:

.TP
.BI "\-f, \-c " conf
Use option settings from the specified configuration file instead of the
default \fIpitrery.conf\fP file. This may be a full path to the file to use,
or the name of a configuration file in \fI/etc/pitrery\fP.

.TP
.B \-l
List available configuration files in the default directory.

.TP
.B \-V
Display the \fBpitrery\fP version and exit.

.TP
.B \-?
Output the brief help summarising the above options.


.SH ACTIONS
Each of the actions also has its own individual set of options, which must
be passed after the action name.


.SS list
The \fBlist\fP action is used to obtain information about the available base
backups which can be used to begin a recovery operation.

.TP 4
.B SYNOPSIS
.B pitrery
.RI [ options ]
.B list
.RI [ list-options ]
.RI [ [[user@]host:]/path/to/backups ]

.TP 4
.B LIST-OPTIONS
The following options are available for the \fBlist\fP action:

.RS

.TP
.B \-j
Display verbose details in \fBJSON\fP format.

.TP
.B \-v
Display verbose details about the listed base backups.

.TP
.B \-?
Output the brief help summarising the \fBlist\fP options.

.TP
.I [[user@]host:]/path/to/backups
The path to the directory storing the backups. If a host name is
provided, it is used to login to a remote archive with
\fBssh\fP(1). When the host is not given, the lookup is local.

.RE


.SS backup
The \fBbackup\fP action is used to create a new base backup of the present
state of a PostgreSQL cluster. This provides a starting point which WAL
segments can be replayed on top of, so that point in time recovery may be done
for any time after the base backup was completed and for which there is a
sequence of archived WAL segment files up to the desired recovery time.

This must be run on the same machine as the cluster that is to be backed up,
though the base backup may be pushed to a remote machine for storage.

.TP 4
.B SYNOPSIS
.B pitrery
.RI [ options ]
.B backup
.RI [ backup-options ]
.RI [ hostname ]

.TP 4
.B BACKUP-OPTIONS
The following options are available for the \fBbackup\fP action:

.RS

.TP
.BI "\-D " dir
The \fBPGDATA\fP path of the cluster to back up on the local machine. This
will override the \fBPGDATA\fP option from the configuration file.

.TP
.BI "\-s " mode
The storage mode to use for the base backup. The \fImode\fP may be \fBtar\fP
or \fBrsync\fP. This will override the \fBSTORAGE\fP option from the
configuration file.

Using \fBrsync\fP will attempt to hardlink any files that have not changed
since the last base backup was taken. Whether this will actually use less
disk space than a compressed tarball will depend on the size of your cluster
and its pattern of use. It will only try to do this to the most recent base
backup, so it will be equivalent to a plain copy of the \fBPGDATA\fP directory
if the previous backup was not also taken with the rsync method.

.TP
.BI "\-c " compress_bin
The command to use for compressing a base backup taken with the \fBtar\fP
mode. This will override the \fBBACKUP_COMPRESS_BIN\fP option from the
configuration file.

.TP
.BI "\-e " compress_suffix
The suffix to add to a compressed base backup taken with the \fBtar\fP
mode. This will override the \fBBACKUP_COMPRESS_SUFFIX\fP option from the
configuration file.

.TP
.B \-E
Encrypt the file using GnuPG. The list of recipients must given using
the \fB-r\fP option. When this option is used, \fBtar\fP storage method
must be used. This will override the \fBBACKUP_ENCRYPT\fP option from the
configuration file.

.TP
.BI "\-r " keys:...
List of public keys for GPG encryption, it is a colon separeted list
of USER-ID recognized by the \fB--recipient\fP option of \fBgpg\fP(1).
This will override the \fBGPG_ENCRYPT_KEYS\fP option from the
configuration file.

.TP
.B \-t
When naming the backup directory from the stop time of the backup, use
ISO 8601 format. This will override the \fBUSE_ISO8601_TIMESTAMPS\fP
option from the configuration file.

.TP
.B \-T
Timestamp the log messages. This will override the \fBLOG_TIMESTAMP\fP option
from the configuration file (forcing it to be "yes").

.TP
.BI "\-P " psql
The \fBpsql\fP(1) command to use to run SQL statements on the database cluster.
This will override the \fBPGPSQL\fP option from the configuration file.

You should avoid using this to pass extra options to \fBpsql\fP (and instead
use the options below for that). It is mostly for overriding the default
\fBpsql\fP(1) that would otherwise be found in the system \fBPATH\fP.

.TP
.BI "\-h " host
The host address to use to query the database cluster. This may be an IP
address for TCP connection or the path to a unix socket directory for a
local peer connection. It is passed as the \fB\-\-host\fP option to
\fBpsql\fP(1). This will override the \fBPGHOST\fP option from the
configuration file.

.TP
.BI "\-p " port
The port to use to query the database cluster. It is passed as the
\fB\-\-port\fP option to \fBpsql\fP(1). This will override the \fBPGPORT\fP
option from the configuration file.

.TP
.BI "\-U " name
The user name to use to query the database cluster. It is passed as
the \fB\-\-username\fP option to \fBpsql\fP(1). This will override
the \fBPGUSER\fP option from the configuration file. This role must
be a superuser or have the replication attribute to execute
\fBpg_start_backup\fP() on the cluster.

.TP
.BI "\-d " database
The database to use when querying the cluster. It is passed as the
\fB\-\-dbname\fP option to \fBpsql\fP(1). This will override the
\fBPGDATABASE\fP option from the configuration file. Note that this does not
influence what is included in the base backup, point in time recovery is
always for the entire cluster.

.TP
.B \-?
Output the brief help summarising the \fBbackup\fP options.

.TP
.I [[user@]host:]/path/to/backups
The path to the directory storing the backups. If a host name is
provided, it is used to login to a remote archive with
\fBssh\fP(1). When the host is not given, the lookup is local.

.RE


.SS restore
The \fBrestore\fP action is used to select and retrieve the files needed to
begin recovery of a cluster to a particular state, either to the most recently
archived state or a point in time between the oldest base backup and the most
recent WAL segment that is available to be replayed.

It will create a new \fBPGDATA\fP tree from the archived data with a minimal
\fIrecovery.conf\fP ready to begin recovery operations. It may also place a
copy of the \fBpostgres\fP configuration files from the time that the base
backup was made in \fIPGDATA/restored_config_files\fP if they did not exist in
the \fBPGDATA\fP directory at that time.

It will create a new \fBPGDATA\fP tree from the archived data with a minimal set
of recovery configuration keys in \fIpostgresql.conf\fP with a "*.signal" file
(PG>=12) or a minimal \fIrecovery.conf\fP (PG<=11) ready to begin
recovery operations.
It may also place a copy of the \fBpostgres\fP configuration files from the time
that the base backup was made in  \fIPGDATA/restored_config_files\fP if they did
not exist in the \fBPGDATA\fP directory at that time.


.TP 4
.B SYNOPSIS
.B pitrery
.RI [ options ]
.B restore
.RI [ restore-options ]
.RI [ [[user@]host:]/path/to/backups ]

.TP 4
.B RESTORE-OPTIONS
The following options are available for the \fBrestore\fP action:

.RS

.TP
.BI "\-D " dir
The \fBPGDATA\fP path of the cluster on the local machine, that is to be
repopulated ready for recovery. This directory will be created if it does not
already exist, but must be empty if it does (unless the \fB\-R\fP option is
used to force overwriting it). This will override the \fBPGDATA\fP option
from the configuration file.

.TP
.BI "\-x " dir
The directory where WAL segment files will be placed if you wish to
keep those outside of the \fBPGDATA\fP tree. If specified this will
create \fIPGDATA/pg_wal\fP (or \fIPGDATA/pg_xlog\fP as of PostgreSQL
9.6 or less) as a symlink to \fIdir\fP rather than as a directory in
its own right. This will override the \fBPGWAL\fP option from the
configuration file.

.TP
.BI "\-d " date
The initial \fIrecovery_target_time\fP to place in
\fIpostgresql.conf\fP (PG>=12) or in \fIrecovery.conf\fP (PG<=11) which is
the first point in time that replaying the WAL segment files will pause at.
The canonical form of the \fIdate\fP string is:

.nh
.nf
  \fIYYYY\-MM\-DD HH:MM:SS\fP [\fI(+|\-)XXXX\fP]
.fi
.hy

where \fIXXXX\fP is the optional timezone offset, however the \fIdate\fP may be
specified here in any form that \fBdate\fP(1) on your system will recognise,
including the relative date strings such as '1\ hour\ ago' which GNU \fBdate\fP
accepts.

This cannot be earlier than the oldest archived base backup, and can only be
restored to if all the WAL segment files from the nearest base backup to that
time are available and uncorrupted.

.TP
.BI "\-O " user
The user which should be set as the owner of the restored files if
\fBpitrery\fP is run as root. This will override the \fBPGOWNER\fP option
from the configuration file.

.TP
.BI "\-t " tblspc:dir
Change the target directory of tablespace \fItblspc\fP to \fIdir\fP. This
option may be used as many times as required if multiple tablespaces need to
to relocated.

.TP
.B \-n
Do a dry run of the restore, showing information about what it would do but
stopping before actually making any changes to \fBPGDATA\fP.

.TP
.B \-R
Overwrite destination directories. By default the \fBrecovery\fP action will
refuse to proceed if any of the destination directories are not empty. Even
with this option it will still refuse to proceed if a \fIpostmaster.pid\fP
file is present, since attempting a restore into directory that a running
\fBpostgres\fP instance is using is likely to Go Very Badly.

.TP
.BI "\-c " uncompress_bin
The command to use for uncompressing a base backup taken with the \fBtar\fP
mode. This will override \fBBACKUP_UNCOMPRESS_BIN\fP option from the
configuration file.

.TP
.BI "\-e " compress_suffix
The file suffix to expect (e.g., gz, bz2, xz) for a compressed base backup
taken with the \fBtar\fP mode. This will override the
\fBBACKUP_COMPRESS_SUFFIX\fP option from the configuration file.

.TP
.BI "\-m " restore_mode
Restore either in \fBstandby\fP or \fBrecovery\fP mode, which will create
respectivily a \fIstandby.signal\fP or \fPrecovery.signal\fP file. This will
override the \fBRESTORE_MODE\fP option from the configuration file.

.TP
.BI "\-r " command
The command line to use in the \fIrestore_command\fP option written in the
\fIpostgresql.conf\fP file (PG>=12) or of the generated \fIrecovery.conf\fP
file (PG<=11). This will override the \fBRESTORE_COMMAND\fP option from the
configuration file. The default is to use \fBrestore_wal\fP(1).

.TP
.BI "\-C " config
The configuration file to use for \fBrestore_wal\fP(1) if
\fBRESTORE_COMMAND\fP was not explicitly specified on either the command line
or in the configuration file.

.TP
.BI "\-y " command
The command to use to set the \fIrecover_end_command\fP option written in the
\fIpostgresql.conf\fP file (PG>=12) or of the generated \fIrecovery.conf\fP
file (PG<=11). This will override the \fBRECOVERY_END_COMMAND\fP option from
the configuration file.

.TP
.B \-T
Timestamp the log messages. This will override the \fBLOG_TIMESTAMP\fP option
from the configuration file (forcing it to be "yes").

.TP
.B \-?
Output the brief help summarising the \fBrestore\fP options.

.TP
.I [[user@]host:]/path/to/backups
The path to the directory storing the backups. If a host name is
provided, it is used to login to a remote archive with
\fBssh\fP(1). When the host is not given, the lookup is local.

.RE


.SS purge
The \fBpurge\fP action is used to perform an orderly expiry of old archived
data that you no longer wish to retain. It will remove both base backups and
any archived WAL segment files that would no longer be usable with just the
base backups that remain. (It will not remove any archived WAL segment files
if there are no base backups at all though).

Expiry of backups can be based on the maximum number of them that you wish to
keep, the maximum age of them that you wish to keep, or a combination of both
where they will only be removed if they exceed both the age limit and the
limit on the number of backups to retain. This can avoid accidentally removing
all the existing backups if all of them are older than the maximum age.

.TP 4
.B SYNOPSIS
.B pitrery
.RI [ options ]
.B purge
.RI [ purge-options ]
.RI [ [[user@]host:]/path/to/backups ]

.TP 4
.B PURGE-OPTIONS
The following options are available for the \fBpurge\fP action:

.RS

.TP
.BI "\-a " [[user@]host:]/dir

CHANGELOG
pitrery 3.3
===========

  * [Bugfix] pitrery restore: could not get the ending timestamp (#138)

pitrery 3.2
===========

  * Pitrery version 3.2 is compatible with PostgreSQL version 13.

  * Add CI for local tests (see CONTRIBUTING.md).

  * Abort check if backup_timestamp is not available (#130)

pitrery 3.1
===========

  * Preserve "xlog" scripts as symlinks in debian and rpm packages #97
    Please update your "archive_command" and "restore_command" parameters to use
    the new scripts "archive_wal" and "restore_wal".

  * Report "qw" update to archive_wal and restore_wal #94, #110 (thanks pgstef)
    Please update if you are using bash version < 4.2.

  * Release process review

pitrery 3.0
===========

  * Pitrery version 3.0 is compatible with PostgreSQL version 12.
    PostgreSQL in this version has changed how recovery configuration is managed.

    There is be no more "recovery.conf" file:
      - The recovery configuration keys are read from the postgresql.conf file.
        Upon restoration, recovery.conf settings are written by pitrery at the
        end of the postgresql.conf file.
      - A "recovery.signal" or "standby.signal" will be used by the restored
        cluster to know what action to take
        A new option is declared: RESTORE_MODE (or the "-m" switch). For
        restoration, it must be set either to "recovery" or "standby".
    - Multiple conflicting recovery_target* specifications are not allowed.
      PostgreSQL will make the check on startup.
    - We now advance to the latest timeline by default.

  * Since PostgreSQL version 10, "xlog" has been renamed to "wal". Make
    this change in pitrery:
    - "archive_xlog" script is renamed to "archive_wal"
    - "restore_xlog" script is renamed to "restore_wal"
    - configuration key PGXLOG is renamed to PGWAL
    - the WAL archive directory ARCHIVE_DIR is set by default to:
      "$BACKUP_DIR/archived_wal"

pitrery 2.4
===========

  * New maintainer: Thibaut Madelaine.

  * Warn about untested major version on restore.


pitrery 2.3
===========

  * Add progress information when checking for missing archived WAL files

  * Support different WAL sizes in check, introduced in PostgreSQL 11

  * Bugfixes in check command


pitrery 2.2
===========

  * Change the command line switch used to pass a configuration file to pitrery
    from -c to -f.

  * Optionnaly encrypt tar backups and archived WAL files.

  * Store the commandline executed and the configuration file inside the backup


pitrery 2.1
===========

  * Restrict permissions of backup directories to the user that created them

  * Add options to the check action to verify backup policy and archived WAL
    files

  * Add RSYNC_WHOLEFILE and RSYNC_BWLIMIT configuration options to
    tweak the behaviour of rsync when performing base backups.



pitrery 2.0
===========

  * Regroup all action into the main script

  * Change the sed commands in the Makefile to have the script work
    without modifications

  * Command line switches and options to specify if the backup is
    local, the user, host and target directory, are now merged into a
    SSH style syntax: `[[user@]host:]/path`. Not providing a host
    tells pitrery the backups are local.

  * Rename the default configuration file from `pitr.conf` to `pitrery.conf`

  * Comment out all parameters in the configuration file so that they
    do not interfere

  * Remove the BACKUP_LABEL option, subdirectory and -l switch. They
    were not used.

  * Support for PostgreSQL 10


pitrery 1.13
============

  * Fix tablespaces links being overwritten by postmaster on 9.5 due
    to tablespace_map. This conflicted with our tablespace relocation

  * Fix the purge of WAL file that would remove all archives if the
    stop file is missing

  * Fix backups failing due to incorrect parsing of the stop time
    when iso8601 timestamps are used


pitrery 1.12
============

  * New -m option for archive_xlog to chmod the archived file to the
    specified octal permission bits

  * Fix the restore so that the backup_label is restore whan used with
    PostgreSQL 9.6

  * Remove support of deprecated compression parameters, renamed in 1.9


pitrery 1.11
============

  * New configure action to create a configuration file from the
    command line.

  * New check action to test configuration files.

  * Do not backup replication slots state files and produce an SQL
    file to recreate them after restore.

  * Add a dry run option (-N) for purge.

  * Improve the quoting and ensure unusual characters are handled
    properly.

  * Add manpages for the user facing commands, pitrery, archive_xlog
    and restore_xlog.

  * Relax restrictions on the target date format on restore. It was
    meant to have a working format common to all free unices
    platforms, but the user can just adapt the command to its
    system. So now one can do run: pitrery restore -d '1 hour ago', or
    anything the local date command can parse.

  * New ARCHIVE_OVERWRITE configuration parameter along with -O option
    (do not overwrite) for archive_xlog to check if the target files
    exists before archiving, when set to "no" or -O is on the command
    line.

  * New ARCHIVE_CHECK (-H) and ARCHIVE_FLUSH (-F) for archive_xlog. The
    first one check md5 sum of the archived WAL file, the second force
    a flush to disk.

  * Make use of the new backup API of PostgreSQL 9.6

  * Optionally name backups dir with ISO 8601 timestamps, with config
    option USE_ISO8601_TIMESTAMPS and command line switch -t for the
    backup action.

  * French translation of the documentation.

  * Bugfixes and better handling of unusual characters in paths


pitrery 1.10
============

  * Fix issue in restore_xlog not using SSH user and host found in the
    configuration file, thus preventing restore to work out of the
    box.

  * Ensure archive_xlog create the target directory on remote hosts


pitrery.bash
#!@BASH@
#
# Copyright 2011-2017 Nicolas Thauvin. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 
#  1. Redistributions of source code must retain the above copyright
#     notice, this list of conditions and the following disclaimer.
#  2. Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY EXPRESS OR
# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#

config_dir="@SYSCONFDIR@"
config=pitr.conf
scripts_dir="@LIBDIR@"
list_configs="no"

usage() {
    echo "usage: `basename $0` [options] action [args]"
    echo "options:"
    echo "    -c file      Path to the configuration file"
    echo "    -n           Show the command instead of executing it"
    echo "    -l           List configuration files in the default directory"
    echo "    -V           Display the version and exit"
    echo "    -?           Print help"
    echo
    echo "actions:"
    echo "    list"
    echo "    backup"
    echo "    restore"
    echo "    purge"
    echo "    check"
    echo "    configure"
    echo
    exit $1
}

error() {
    echo "ERROR: $*" 1>&2
    exit 1
}

info() {
    echo "INFO: $*"
}

while getopts "c:nlV?" opt; do
    case $opt in
	c) config=$OPTARG;;
	n) dry_run="echo";;
	l) list_configs="yes";;
	V) echo "pitrery @VERSION@"; exit 0;;
	'?') usage 1;;
	*) error "error while processing options";;
    esac
done

# Ensure failed globs will be empty, not left containing the literal glob pattern
shopt -s nullglob

# List the configuration files and exit
if [ "$list_configs" = "yes" ]; then
    info "listing configuration files in $config_dir"
    for x in "$config_dir"/*.conf; do
	basename -- "$x" .conf
    done
    exit 0
fi

if (( $# < 1 )); then
    echo "ERROR: missing action" 1>&2
    usage 1
fi

action=${@:$OPTIND:1}
OPTIND=$(( $OPTIND + 1 ))

# Check if the config option is a path or just a name in the
# configuration directory.  Prepend the configuration directory and
# .conf when needed.
if [[ $config != */* ]]; then
    config="$config_dir/$(basename -- "$config" .conf).conf"
fi

load_config() {
    # Load the configuration file
    if [ -f "$config" ]; then
	. "$config"
    else
	error "cannot access configuration file: $config"
    fi
}

select_cmd() {
    # Find the command to run
    cmd=${scripts_dir:-.}/$1
    if [ ! -f "$cmd" ] || [ ! -x "$cmd" ]; then
	error "command '$cmd' is unusable"
    fi
}

run_cmd() {
    # Append the remote hostname option if provided
    if [ "$BACKUP_IS_LOCAL" != "yes" ]; then
	if [ -n "$1" ]; then
	    opts+=( "$1" )
	elif [ -n "$BACKUP_HOST" ]; then
	    opts+=( "$BACKUP_HOST" )
	else
	    error "remote backup hostname not specified"
	fi
    fi

    # Run the command
    $dry_run "$cmd" "${opts[@]}"
    exit $?
}

opts=()
case $action in
    list)
	load_config
	select_cmd "list_pitr"

	# Parse args after action: they should take precedence over the configuration
	while getopts "Lu:b:l:v?" arg 2>/dev/null; do
	    case $arg in
		L) BACKUP_IS_LOCAL="yes";;
		u) BACKUP_USER=$OPTARG;;
		b) BACKUP_DIR=$OPTARG;;
		l) BACKUP_LABEL=$OPTARG;;
		v) VERBOSE="yes";;
		'?') "$cmd" '-?'; exit $?;;
	    esac
	done

	# Add relevant options coming from the configuration
	[ "$BACKUP_IS_LOCAL" = "yes" ]	&& opts+=( "-L" )
	[ -n "$BACKUP_USER" ]		&& opts+=( "-u" "$BACKUP_USER" )
	[ -n "$BACKUP_DIR" ]		&& opts+=( "-b" "$BACKUP_DIR" )
	[ -n "$BACKUP_LABEL" ]		&& opts+=( "-l" "$BACKUP_LABEL" )
	[ "$VERBOSE" = "yes" ]		&& opts+=( "-v" )

	run_cmd "${@:$OPTIND:1}"
	;;

    backup)
	load_config
	select_cmd "backup_pitr"

	# Parse args after action: they should take precedence over the configuration
	while getopts "Lb:l:u:D:s:P:h:p:U:d:c:e:tT?" arg 2>/dev/null; do
	    case $arg in
		L) BACKUP_IS_LOCAL="yes";;
		b) BACKUP_DIR=$OPTARG;;
		l) BACKUP_LABEL=$OPTARG;;
		u) BACKUP_USER=$OPTARG;;
		D) PGDATA=$OPTARG;;
		s) STORAGE=$OPTARG;;
		P) PGPSQL=$OPTARG;;
		h) PGHOST=$OPTARG;;
		p) PGPORT=$OPTARG;;
		U) PGUSER=$OPTARG;;
		d) PGDATABASE=$OPTARG;;
		c) BACKUP_COMPRESS_BIN=$OPTARG;;
		e) BACKUP_COMPRESS_SUFFIX=$OPTARG;;
                t) USE_ISO8601_TIMESTAMPS="yes";;
		T) LOG_TIMESTAMP="yes";;

		'?') "$cmd" '-?'; exit $?;;
	    esac
	done

	# Add relevant options coming from the configuration
	[ "$BACKUP_IS_LOCAL" = "yes" ]	      && opts+=( "-L" )
	[ -n "$BACKUP_DIR" ]		      && opts+=( "-b" "$BACKUP_DIR" )
	[ -n "$BACKUP_LABEL" ]		      && opts+=( "-l" "$BACKUP_LABEL" )
	[ -n "$BACKUP_USER" ]		      && opts+=( "-u" "$BACKUP_USER" )
	[ -n "$PGDATA" ]		      && opts+=( "-D" "$PGDATA" )
	[ -n "$STORAGE" ]		      && opts+=( "-s" "$STORAGE" )
	[ -n "$PGPSQL" ]		      && opts+=( "-P" "$PGPSQL" )
	[ -n "$PGHOST" ]		      && opts+=( "-h" "$PGHOST" )
	[ -n "$PGPORT" ]		      && opts+=( "-p" "$PGPORT" )
	[ -n "$PGUSER" ]		      && opts+=( "-U" "$PGUSER" )
	[ -n "$PGDATABASE" ]		      && opts+=( "-d" "$PGDATABASE" )
	[ -n "$BACKUP_COMPRESS_BIN" ]	      && opts+=( "-c" "$BACKUP_COMPRESS_BIN" )
	[ -n "$BACKUP_COMPRESS_SUFFIX" ]      && opts+=( "-e" "$BACKUP_COMPRESS_SUFFIX" )
        [ "$USE_ISO8601_TIMESTAMPS" = "yes" ] && opts+=( "-t" )
	[ "$LOG_TIMESTAMP" = "yes" ]	      && opts+=( "-T" )

	# If hooks are defined export them
	[ -n "$PRE_BACKUP_COMMAND" ] && export PRE_BACKUP_COMMAND
	[ -n "$POST_BACKUP_COMMAND" ] && export POST_BACKUP_COMMAND

	run_cmd "${@:$OPTIND:1}"
	;;

    restore)
	load_config
	select_cmd "restore_pitr"

	# Parse args after action: they should take precedence over the configuration
	while getopts "Lu:b:l:D:x:d:O:t:nRc:e:r:C:T?" arg 2>/dev/null; do
	    case $arg in
		L) BACKUP_IS_LOCAL="yes";;
		u) BACKUP_USER=$OPTARG;;
		b) BACKUP_DIR=$OPTARG;;
		l) BACKUP_LABEL=$OPTARG;;
		D) PGDATA=$OPTARG;;
		x) PGXLOG=$OPTARG;;
		d) TARGET_DATE=$OPTARG;;
		O) PGOWNER=$OPTARG;;
		t) TBLSPC_RELOC+=( "-t" "$OPTARG" );;
		n) DRY_RUN="yes";;
		R) OVERWRITE="yes";;
		c) BACKUP_UNCOMPRESS_BIN=$OPTARG;;
		e) BACKUP_COMPRESS_SUFFIX=$OPTARG;;
		r) RESTORE_COMMAND=$OPTARG;;
		C) RESTORE_XLOG_CONFIG=$OPTARG;;
		T) LOG_TIMESTAMP="yes";;

		"?") "$cmd" '-?'; exit $?;;
	    esac
	done

	# Add relevant options coming from the configuration
	[ "$BACKUP_IS_LOCAL" = "yes" ]	    && opts+=( "-L" )
	[ -n "$BACKUP_USER" ]		    && opts+=( "-u" "$BACKUP_USER" )
	[ -n "$BACKUP_DIR" ]		    && opts+=( "-b" "$BACKUP_DIR" )
	[ -n "$BACKUP_LABEL" ]		    && opts+=( "-l" "$BACKUP_LABEL" )
	[ -n "$PGDATA" ]		    && opts+=( "-D" "$PGDATA" )
	[ -n "$PGXLOG" ]		    && opts+=( "-x" "$PGXLOG" )
	[ -n "$TARGET_DATE" ]		    && opts+=( "-d" "$TARGET_DATE" )
	[ -n "$PGOWNER" ]		    && opts+=( "-O" "$PGOWNER" )
	(( ${#TBLSPC_RELOC[@]} > 0 ))	    && opts+=( "${TBLSPC_RELOC[@]}" )
	[ "$DRY_RUN" = "yes" ]		    && opts+=( "-n" )
	[ -n "$OVERWRITE" ]		    && opts+=( "-R" )
	[ -n "$BACKUP_UNCOMPRESS_BIN" ]	    && opts+=( "-c" "$BACKUP_UNCOMPRESS_BIN" )
	[ -n "$BACKUP_COMPRESS_SUFFIX" ]    && opts+=( "-e" "$BACKUP_COMPRESS_SUFFIX" )
	[ -n "$RESTORE_COMMAND" ]	    && opts+=( "-r" "$RESTORE_COMMAND" )
	[ "$LOG_TIMESTAMP" = "yes" ]	    && opts+=( "-T" )

	# Pass along the configuration file
	if [ -n "$RESTORE_XLOG_CONFIG" ]; then
	    opts+=( "-C" "$RESTORE_XLOG_CONFIG" )
	elif [ "$config" != "@SYSCONFDIR@/pitr.conf" ]; then
	    opts+=( "-C" "$config" )
	fi

	run_cmd "${@:$OPTIND:1}"
	;;

    purge)
	load_config
	select_cmd "purge_pitr"

	# Parse args after action: they should take precedence over the configuration
	while getopts "Ll:b:u:n:U:X:m:d:NT?" arg 2>/dev/null; do
	    case $arg in
		L) BACKUP_IS_LOCAL="yes";;
		l) BACKUP_LABEL=$OPTARG;;
		b) BACKUP_DIR=$OPTARG;;
		u) BACKUP_USER=$OPTARG;;
		n) ARCHIVE_HOST=$OPTARG;;
		U) ARCHIVE_USER=$OPTARG;;
		X) ARCHIVE_DIR=$OPTARG;;
		m) PURGE_KEEP_COUNT=$OPTARG;;
		d) PURGE_OLDER_THAN=$OPTARG;;
		N) DRY_RUN="yes";;
		T) LOG_TIMESTAMP="yes";;

		"?") "$cmd" '-?'; exit $?;;
	    esac
	done

	# Add relevant options coming from the configuration
	[ "$BACKUP_IS_LOCAL" = "yes" ]	&& opts+=( "-L" )
	[ -n "$BACKUP_DIR" ]		&& opts+=( "-b" "$BACKUP_DIR" )
	[ -n "$BACKUP_LABEL" ]		&& opts+=( "-l" "$BACKUP_LABEL" )
	[ -n "$BACKUP_USER" ]		&& opts+=( "-u" "$BACKUP_USER" )
	[ -n "$ARCHIVE_HOST" ]		&& opts+=( "-n" "$ARCHIVE_HOST" )
	[ -n "$ARCHIVE_USER" ]		&& opts+=( "-U" "$ARCHIVE_USER" )
	[ -n "$ARCHIVE_DIR" ]		&& opts+=( "-X" "$ARCHIVE_DIR" )
	[ -n "$PURGE_KEEP_COUNT" ]	&& opts+=( "-m" "$PURGE_KEEP_COUNT" )
	[ -n "$PURGE_OLDER_THAN" ]	&& opts+=( "-d" "$PURGE_OLDER_THAN" )
	[ "$DRY_RUN" = "yes" ]		&& opts+=( "-N" )
	[ "$LOG_TIMESTAMP" = "yes" ]	&& opts+=( "-T" )

	run_cmd "${@:$OPTIND:1}"
	;;

    check)
	load_config
	select_cmd "check_pitr"

	while getopts "C:?" opt; do
	    case $opt in
		C) PITR_CONFIG=$OPTARG;;
		"?") "$cmd" '-?'; exit $?;;
	    esac
	done

	if [ -n "$PITR_CONFIG" ]; then
	    opts+=( "-C" "$PITR_CONFIG" )
	else
	    opts+=( "-C" "$config" )
	fi

        # Run the command
	$dry_run "$cmd" "${opts[@]}"
	exit $?
	;;

    configure)
	# Here we do not load any configuration so that it does not
	# change the output in the back of the user. This is important
	# for PostgreSQL related environment variables that we do not
	# want to overwrite. This is why the following loop getopts on
	# get just pass in the command line.
	select_cmd "configure_pitr"

	while getopts "o:fCl:s:m:g:D:a:P:h:p:U:d:?" opt; do
	    case $opt in
		o) opts+=( "-o" "$OPTARG" );;
		f) opts+=( "-f" );;
		C) opts+=( "-C" );;
		l) opts+=( "-l" "$OPTARG" );;
		s) opts+=( "-s" "$OPTARG" );;
		m) opts+=( "-m" "$OPTARG" );;
		g) opts+=( "-g" "$OPTARG" );;
		D) opts+=( "-D" "$OPTARG" );;
		a) opts+=( "-a" "$OPTARG" );;

		P) opts+=( "-P" "$OPTARG" );;
		h) opts+=( "-h" "$OPTARG" );;
		p) opts+=( "-p" "$OPTARG" );;
		U) opts+=( "-U" "$OPTARG" );;
		d) opts+=( "-d" "$OPTARG" );;

		"?") "$cmd" '-?'; exit $?;;
	    esac
	done

	[ -n "${@:$OPTIND:1}" ] && opts+=( "${@:$OPTIND:1}" )
	$dry_run "$cmd" "${opts[@]}"
	exit $?
	;;

    *)
	error "unknown action"
	;;
esac
